### PR TITLE
fix: prevent future OpenClaw rewrite usage inflation

### PR DESCRIPTION
**Pull request by EncodeTS**
OpenClaw atomically rewrites active session JSONL files. The previous inode/offset cursor restarted at byte zero after a rewrite, replaying historical assistant usage and repeatedly adding the same cache-write tokens and conversations.

This change prevents future inflation with exact, privacy-safe usage-event deduplication. Stable event IDs are preferred; otherwise a usage-metadata fingerprint with multiset counts preserves legitimately repeated identical events. The parser verifies metadata-only prefix fingerprints before append resume, handles same/new inode rewrites, same-size/larger in-place rewrites, zero/unreliable inode values, compaction/restoration, conservative legacy cursor migration, and concurrent file replacement.

Cursor and hourly state are committed only after queue writes succeed, so a failed enqueue cannot advance past unqueued usage. Exact event identities remain in the per-file cursor for the session lifetime; lossy pruning would allow restored compacted events to be counted again.

Privacy: only event IDs and usage metadata are hashed. Prompts, message bodies, and arbitrary JSONL bytes are never hashed, logged, or persisted.

This PR intentionally does not rewrite historical OpenClaw totals. Legacy aggregate queue rows cannot distinguish replayed events from legitimate equal-token events without event identities, so a generic automatic migration could delete valid usage. Existing affected users require an explicit, evidence-based manual repair workflow. The reporter's local queue was corrected separately; this PR does not encode or replay that one-off operation.

Validation:
- Focused OpenClaw parser tests: 30/30.
- `npm run ci:local`: 1,499/1,499 tests plus validators, architecture guardrails, and dashboard production build.
- Linux runs the full Node suite; native macOS and Windows jobs run the focused OpenClaw parser suite.
- Fresh long-context bug, Standards, and Spec reviews are clean.

No version bump is included. A CLI + macOS + Windows release is required for installed clients to receive future-inflation prevention; it will not automatically rewrite existing historical totals.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved incremental OpenClaw session parsing across atomic rewrites, append-only growth, and path/casing variations, with more reliable cursor reuse and legacy resume/migration.
  * Fixed usage-event deduplication and fingerprint/offset handling to prevent miscounting during partial reads and concurrent updates.
  * Corrected provider identification behavior between OpenClaw sessions and Codex rollout files.
* **Tests**
  * Added end-to-end coverage for incremental parsing, migration, deduplication, fingerprint validation, and edge cases across platforms.
* **Chores**
  * Updated CI to run the new Node-based parser test on macOS and Windows.

### CODE REVIEW

**Comment by EncodeTS** [comment]
Addressing the CodeRabbit `usageEvents` growth nitpick: I intentionally did not cap, time-prune, or replace the ledger with only identities present in the current file. OpenClaw may compact and later rewrite/restore prior events; evicting an identity would make that restored event look new and reintroduce the exact usage inflation this PR fixes. The cursor therefore retains an exact hashed event-identity multiset. It stores usage metadata/IDs only—never prompts or bodies. A compact disk-backed exact ledger could be explored separately, but lossy pruning is not safe here. The follow-up also replaces raw-prefix verification with metadata-only fingerprints and adds regressions for compaction followed by rewrite.

**Comment by EncodeTS** [comment]
@coderabbitai full review

File: src/commands/sync.js [hunk 1]
```diff
@@ -3102,6 +3127,566 @@ async function migrateRolloutCumulativeDeltaBuckets({ cursors, queuePath, rollou
...
+  } catch (e) {
+    if (e?.code !== "ENOENT") uploadState = {};
+  }
+  const offset = Number(uploadState?.offset || 0);
+  const validOffset =
+    Number.isInteger(offset) &&
+    offset >= 0 &&
+    offset <= queue.length &&
+    (offset === 0 || queue[offset - 1] === 0x0a);
+  return { queue, offset, validOffset };
```
[the remark this answers is not in the document]
Comment (EncodeTS) [reply src/commands/sync.js]: Fixed in 09713a9. `readOpenclawQueueState` now distinguishes a missing state file from an existing untrustworthy one: ENOENT keeps the safe offset-0 default, while malformed JSON, non-object JSON, and non-ENOENT read failures set `validOffset=false`. The repair then defers with `openclaw_queue_offset_invalid` before any queue or usage-cursor mutation. Added regressions for malformed JSON, valid non-object JSON, and simulated EACCES; each asserts the queue plus `files`/`hourly` state remain unchanged.

**Comment by EncodeTS** [comment]
Addressed the full-review fingerprint performance finding in `09713a9`. Routine append and one-time legacy adoption still perform the required metadata-only prefix verification/baseline scan, but now retain an ephemeral in-memory `crypto.Hash.copy()` state and extend it with only the tail metadata. Only the existing `{version,eventCount,digest}` remains persisted; no raw bytes, prompt/body data, or hash internals are stored. This removes the redundant final `0..endOffset` scan without weakening same-inode rewrite detection. Range-instrumented tests assert exactly one prefix scan plus one tail scan, and native macOS/Windows CI now runs both parser and repair suites.

**Comment by EncodeTS** [comment]
@coderabbitai full review

File: src/lib/rollout.js [hunk 2]
```diff
@@ -1072,6 +1308,32 @@ async function parseOpenclawSessionFile({
...
+    scanCounts[eventIdentity.key] = nextCount;
+    usageEvents[eventIdentity.key] = eventIdentity.unique
+      ? 1
+      : Math.max(Number(usageEvents[eventIdentity.key] || 0), nextCount);
+    if (
+      eventIdentity.unique
+        ? priorCount > 0 || nextCount > 1
+        : nextCount <= priorCount
+    ) {
+      continue;
```
[the remark this answers is not in the document]
Comment (EncodeTS) [reply src/lib/rollout.js]: The historical repair subsystem was removed in 42d6038. This forward-parser finding remains addressed: malformed, zero-usage, or invalid-timestamp records do not enter usageEvents; strict token-type validation also runs before identity/fingerprint processing, so a later corrected record with the same stable ID counts exactly once.

File: src/commands/sync.js [hunk 3]
```diff
@@ -3102,6 +3127,573 @@ async function migrateRolloutCumulativeDeltaBuckets({ cursors, queuePath, rollou
...
+  const markerPath = openclawRepairMarkerPath(queuePath);
+  const pendingPlan = await readJson(markerPath);
+  if (prior && !(typeof prior === "object" && prior.status === "deferred")) {
+    await fs.rm(markerPath, { force: true });
+    return false;
+  }
+  if (pendingPlan) {
+    const corrections = openclawRepairPlanCorrections(pendingPlan);
+    const queueState = await readOpenclawQueueState(queuePath, queueStatePath);
+    if (!corrections || !queueState.validOffset) {
```
[the remark this answers is not in the document]
Comment (EncodeTS) [reply src/commands/sync.js]: The automatic OpenClaw historical repair subsystem has been removed entirely in 42d6038. There is no repair marker, queue-state rewrite, migration key, or correction application path left to validate. The PR is now forward-prevention only because legacy aggregate rows cannot prove which totals are replayed versus legitimate.

File: src/commands/sync.js [hunk 4]
```diff
@@ -3102,6 +3127,573 @@ async function migrateRolloutCumulativeDeltaBuckets({ cursors, queuePath, rollou
...
+  const trackedPaths = new Set();
+  for (const [filePath, cursor] of Object.entries(cursors.files || {})) {
+    if (isOpenclawTrackedUsagePath(filePath, cursor, roots)) {
+      trackedPaths.add(filePath);
+    }
+  }
+  for (const entry of Array.isArray(sessionFiles) ? sessionFiles : []) {
+    const filePath = typeof entry === "string" ? entry : entry?.path;
+    if (typeof filePath === "string" && filePath) trackedPaths.add(filePath);
+  }
```
[the remark this answers is not in the document]
Comment (EncodeTS) [reply src/commands/sync.js]: The automatic OpenClaw historical repair subsystem has been removed entirely in 42d6038, including tracked repair paths and contributor counting. Windows path normalization remains only in the forward parser cursor key, where aliases are canonicalized before cursor lookup.

**Comment by EncodeTS** [comment]
@coderabbitai full review

**Comment by EncodeTS** [comment]
@coderabbitai full review

**Comment by EncodeTS** [comment]
@coderabbitai full review

**Comment by EncodeTS** [comment]
On the final CodeRabbit prefix-rescan nitpick: the O(prefix) metadata scan is intentional. Persisted SHA-256 internal state is not safely serializable across processes, and inode/size/mtime are not reliable enough to detect same-inode same-size/larger rewrites. The exact event ledger prevents replay but cannot prove the prior byte range is unchanged. Re-scanning usage metadata (never prompt/body bytes in persisted state) is therefore the correctness-preserving check; capping/pruning would reintroduce inflation after compaction/restoration, and transcript rotation is owned by OpenClaw rather than TokenTracker. No change made for this trivial tradeoff.
